### ISSUE TITLE
[async] [filesystem] Incompatible global `ErrorCallback` interfaces

### ISSUE THREAD

**Issue by jvilk** [DefinitelyTyped/DefinitelyTyped#12942]
- [x] I tried using the latest `filesystem/filesystem.d.ts` and `async/async.d.ts` files in this repo and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] I want to talk about `filesystem/filesystem.d.ts` and `async/async.d.ts`.
  - The authors of filesystem is cc/ @kontan
  - The authors of async are @borisyankov @kern0 @Penryn @fenying 

## Issue

Both `filesystem` and `async` typings define incompatible and *global* `ErrorCallback` interfaces.

`filesystem.d.ts`:

```typescript
/**
 * When an error occurs, the following callback is made.
 */
interface ErrorCallback {
    (err:DOMError):void;
}
```

`async.d.ts`:

```typescript
interface ErrorCallback<T> { (err?: T): void; }
```

This incompatibility recently (and silently!) broke the build process [for my BrowserFS library](https://github.com/jvilk/BrowserFS/issues/165).

## Suggested Fix

Follow best practices and put all of these currently-global interfaces in these typings into namespaces, like we do for NodeJS and other modules. This may break existing TypeScript codebases, though, since the \@types mechanism has no notion of a major typings version change and folks may be depending on the global types. I argue that it is worth this problem, since it will prevent future issues.

A simpler and *much worse* fix is to "kick the can down the road" and rename one of the `ErrorCallback` interfaces. If we apply this "quick fix", though, then I am sure that this problem will occur again in the future!

Thoughts? If the typing authors do not want to undertake this large change, I can take a stab at it.

**Comment by jvilk** [comment]
(Note that these interfaces coexisted peacefully for a long time because they coincidentally defined compatible types.)

**Comment by Penryn** [comment]
It's caught up to us, huh? I personally don't like having global `ErrorCallback`s, however I am guilty of contributing to this problem for keeping it in `async.d.ts`.

I vote for eventual namespacing despite the heavy cost of breaking everyone's repos.

However, `async.d.ts` should at least be amended so ErrorCallback is not generic, because right now it is wrong:
```javascript
interface ErrorCallback{
    (error?: Error): void;
}
```
Would this be sufficient to start?

**Comment by jvilk** [comment]
@Penryn Yup! That is actually the definition from the previous revision, before it was genericized:

```typescript
interface ErrorCallback { (err?: Error): void; }
``` 

How is it wrong to be generic, btw? When using async, you can pass whatever you want into the library as an "error" object.

**Comment by Penryn** [comment]
You're technically correct (the best kind). I haven't run into situations requiring code to return anything other than an `Error`-derived object, but after experimenting, keeping the generic version could work fine for `async.d.ts`. But that leaves the original situation.

I was going to suggest that `filesystem.d.ts` adopt the `ErrorCallback` of `async.d.ts`, but it appears the [`error` is not optional](https://dev.w3.org/2009/dap/file-system/file-dir-sys.html#the-errorcallback-interface), and therefore these two types are incompatible.

I'm out of ideas. Maybe we should just namespace each.

**Comment by jvilk** [comment]
Yeah, namespacing is the correct technical solution. Will people hate us if we refactor the typings like this?

> I was going to suggest that filesystem.d.ts adopt the ErrorCallback of async.d.ts, but it appears the error is not optional, and therefore these two types are incompatible.

[Actually, they should be compatible due to function parameter bivariance (ctrl+f on this page for it)](https://www.typescriptlang.org/docs/handbook/type-compatibility.html). The following code compiles:

```typescript
interface ErrorCallback { (err?: Error): void; }
interface ErrorCallback { (err: Error): void; }
```

**Comment by Penryn** [comment]
> Will people hate us if we refactor the typings with namespaces?
Probably, but it'd mainly affect globally declared types that people actually use. This is Javascript world, where the code you wrote five hours ago is deprecated because frameworks. So let's just do it right and namespace it.

What you point out about bivariance makes sense--I see it's what let your code compile before `async.d.ts` went generic.

Ok so, namespacing: yes.
Aligning `ErrorCallback`s: no? Maybe? What's your vote?

**Comment by jvilk** [comment]
If we namespace, we shouldn't need to align `ErrorCallback` types.

**Comment by fenying** [comment]
I don't know why ***async.d.ts*** was defined as global definitions, and today it's useless to talk about it. 

Now, the exactly right way to solve this problem once for all, is changing both **filesystem.d.ts** and **async.d.ts** into modular definitions, but it could break almost all repos using these two **.d.ts** files. That could be a great expense. And also for this, in previous editions, nobody did change it into a modular definition, just to make least changes and keep the backward compatibilities.

@Penryn

Thus, I think, shall we create a tag version as a compatible version, and then change these files **.d.ts** files into modular defintions in the future version? As the way `@types/node@4` and `@types/node@6` works. So that people can solve the compatibility problem simply by using a compatible version.

**Comment by Penryn** [comment]
@fenying: After some thought, I'm just going to update the *types-2.0* branch. We'll leave the *master * branch intact to preserve old repos, but there's no need to create two new versions since the underlying async API hasn't changed.

**Comment by jvilk** [comment]
Where did we land on this? I just recently had this issue bite me in a recent repository, and thought that `async` would've been namespaced appropriately now (unless I misunderstood @Penryn's comment). I'm importing the types using NPM, which I suspect uses the types-2.0 branch. (I haven't been keeping up with how types are being managed now.)

If not, maybe I should change the filesystem typings and bring it up a major version to avoid making people angry. It's not a versioned library, it's a (*deprecated*) browser feature. Is @kontan, the maintainer, active anymore?

**Comment by Penryn** [comment]
Really sorry but life and work are quite busy for me at the moment. May have to step down as maintainer for a bit as other obligations have my attention at the moment.

**Comment by orta** [comment]
Hi thread, we're moving DefinitelyTyped to use [GitHub Discussions](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/53377) for conversations the `@types` modules in DefinitelyTyped.

To help with the transition, we're closing all issues which haven't had activity in the last 6 months, which includes this issue. If you think closing this issue is a mistake, please pop into the [TypeScript Community Discord](https://discord.gg/typescript) and mention the issue in the `definitely-typed` channel.
